Menu control position according Drop mirroring

leossantiago opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

Right now, Menu doesn't have a proper way to update Menu Control button when `Drop` is mirrored:

![Issue3540](https://user-images.githubusercontent.com/47114840/104136139-27dbe900-5373-11eb-86a1-8a23df14a1f1.gif)
(this can be reproduced on https://storybook.grommet.io/?path=/story/visualizations-list--action)

Even though it's on List stories, this problem is caused by `Menu`.
When using a `Menu` , it renders this sequence of components: Menu -> DropButton -> Button ->  Drop -> DropContainer. 

While Menu control button (three dots) position changes according to Menu `align` prop (which defaults to `top: 'top'` vertically, but can be set as `bottom: 'bottom'`), `DropContainer` calculates remaining space on view and forces a mirrored position if there's not enough space to allocate the container. And since those actions are not combined, it results on the problem showed above: a mirrored Drop with Menu-control button on top.

#### What does this PR do?

Fixes #3540 by keeping track of `DropContainer` align on `Menu`, forcing a change on control button position when Drop is mirrored.

![Issue3540_2](https://user-images.githubusercontent.com/47114840/104136479-34f9d780-5375-11eb-914f-9c582f613768.gif)

setState function is passed down to DropContainer, and it changes according to its align prop whenever there's an update to DropContainer style.

This PR also fixes a typing problem on `List` `pad` prop, where `'none'` wasn't being recognized as a valid type (`pad={{ left: 'small', right: 'none' }}` on example below).

#### Where should the reviewer start?

src/js/components/Menu/Menu.js
src/js/components/Drop/DropContainer.js

#### What testing has been done on this PR?

Manual testing using code below.

#### How should this be manually tested?

```js
  <Grommet theme={grommet}>
    <Box pad="large">
      <List
        data={data.slice(0, 10)}
        pad={{ left: 'small', right: 'none' }}
        action={(item, index) => {
          return (
            <Menu
              dropAlign={{ top: 'top', left: 'left' }}
              key={index}
              icon={<More />}
              hoverIndicator
              items={[{ label: 'one' }]}
            />
          );
        }}
      />
    </Box>
  </Grommet>
```

#### Any background context you want to provide?

#### What are the relevant issues?

Issue #3540 

#### Screenshots (if appropriate)

GIFs above.

#### Do the grommet docs need to be updated?

No.

#### Should this PR be mentioned in the release notes?

#### Is this change backwards compatible or is it a breaking change?

Breaking change if you consider position change when mirrored. Otherwise, backwards compatible.

ShimiSun commented:
@leossantiago we missed you buddy!! I'll try to review the PR today!

ShimiSun commented on src/js/components/List/stories/Action.js:
```diff
@@ -17,7 +17,8 @@ export const Action = () => (
         data={data.slice(0, 10)}
-        pad={{ left: 'small', right: 'none' }}
+        pad={{ left: 'large', right: 'small' }}
+        // pad={{ left: 'small' }}
```
Why is this comment needed?

leossantiago replied:
Thanks for pointing that! I was using this story as a playground and didn't realize this test was added to the commit. I'll remove it on the next commit. :smile:

ShimiSun commented:
> Breaking change if you consider position change when mirrored. Otherwise, backward compatible.

Good point, I think that it would make sense for Menu to open in the way you have suggested. I'll take a few hours to think of counter examples.

taysea commented on src/js/components/Menu/Menu.js:
```diff
@@ -275,7 +281,9 @@ const Menu = forwardRef((props, ref) => {
           >
             <ContainerBox background={dropBackground || theme.menu.background}>
-              {align.top === 'top' ? controlMirror : undefined}
+              {menuMirrorReference === 'top' && align.top === 'top'
```
Thinking functionality looks great. Just wondering if this variable could have a more clear name.

maybe, `alignControlMirror` and `setAlignControlMirror` ?

leossantiago replied:
I'm open to any suggestions. I must confess I spent a couple minutes renaming and trying new names before committing so it would be more clear haha. I'll be pushing a new commit to remove a comment that wasn't supposed to be there and guess  we'll go with `alignControlMirror` and `setAlignControlMirror`. :smile:

leossantiago commented:
@ShimiSun feels good to be back. Missed this community as well :smile: 
I'll be addressing reviews soon, and we can discuss counter examples when you come up with any.

taysea commented on src/js/components/Drop/DropContainer.js:
```diff
@@ -62,6 +63,13 @@ const DropContainer = forwardRef(
     useEffect(() => {
+      const updateMenuMirrorReference = () => {
+        const styleCurrent = (ref || dropRef).current.style;
+        const sideToAlign = styleCurrent.top !== '' ? 'top' : 'bottom';
```
soft opinion for variable name.

```suggestion
        const alignControl = styleCurrent.top !== '' ? 'top' : 'bottom';
```

leossantiago commented:
Addressed comments :tada:

ericsoderberghp commented on src/js/components/Drop/DropContainer.js:
```diff
@@ -37,6 +37,7 @@ const DropContainer = forwardRef(
   (
     {
       align = defaultAlign,
+      setAlignControlMirror,
```
Can we call this `onAlign`? I think that will be more consistent and generic, caller is notified when the actual alignment changes.

ericsoderberghp reviewed (commented):
Overall, this looks good to me. Just one small suggestion for better consistency.

leossantiago commented:
@ericsoderberghp updated props name on `DropContainer` and `DropButton` but kept it as it is on `Menu` so the state stays self-explanatory since `Menu` is the one handling the change of position. What do you think?

Also, on `DropContainer` I kept the whole updating function as `updateAlignControlMirror`. I'd like your opinion on that as well!

ericsoderberghp commented on src/js/components/Drop/DropContainer.js:
```diff
@@ -62,6 +63,13 @@ const DropContainer = forwardRef(
     ]);
     const dropRef = useRef();
     useEffect(() => {
+      const updateAlignControlMirror = () => {
```
Agreed with leaving Menu as is. But, can we call this function something like "notifyAlign", since DropContainer doesn't know anything about a "control mirror".

leossantiago replied:
Agreed! Addressed!

ericsoderberghp commented on src/js/components/Drop/DropContainer.js:
```diff
@@ -225,6 +233,7 @@ const DropContainer = forwardRef(
             container.style.maxHeight = `${maxHeight}px`;
           }
         }
+        updateAlignControlMirror();
```
Can we move the `if (onAlign)` check here, since it won't do anything unless the caller passes it in?

leossantiago replied:
Agreed! Addressed!

ericsoderberghp reviewed (commented):
One more small naming suggestion.

ShimiSun commented:
@leossantiago hi buddy, I would love to get this PR to the upcoming release, any chance you can address the comments today? Thank you for working on it 💟

leossantiago commented:
Hey @ShimiSun! Sorry for not addressing that earlier. It was already late night around here when you tagged me here 😅 
Addressed Eric's suggestions. Hope it's still on time. Ping me if something else should be changed! 🎉

ShimiSun commented:
Thank you @leossantiago! I'm glad this fix will be available on the release. 💟
